crypto/src/bulletproofs.rs

```diff
@@ -0,0 +1,218 @@
+
+    fn add(self, rhs: PedersenCommitment) -> PedersenCommitment {
+        PedersenCommitment {
+            point: self.point + rhs.point,
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl ops::Sub<PedersenCommitment> for PedersenCommitment {
+    type Output = PedersenCommitment;
+
+    fn sub(self, rhs: PedersenCommitment) -> PedersenCommitment {
+        PedersenCommitment {
+            point: self.point - rhs.point,
+            bytes: OnceCell::new(),
+        }
+    }
+}
+
+impl AsRef<[u8]> for PedersenCommitment {
+    fn as_ref(&self) -> &[u8] {
+        self.bytes.get_or_init(|| self.point.compress().to_bytes())
+    }
+}
+
+impl ToFromBytes for PedersenCommitment {
+    fn from_bytes(bytes: &[u8]) -> Result<Self, signature::Error> {
+        if bytes.len() != 32 {
+            return Err(signature::Error::new());
+        }
+        let point = CompressedRistretto::from_slice(bytes);
+        let decompressed_point = point.decompress().ok_or_else(signature::Error::new)?;
+
+        Ok(PedersenCommitment {
+            point: decompressed_point,
+            bytes: OnceCell::new(),
+        })
+    }
+}
+
+impl Serialize for PedersenCommitment {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let bytes = self.as_ref();
+        serializer.serialize_bytes(bytes)
+    }
+}
+
+impl<'de> Deserialize<'de> for PedersenCommitment {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: de::Deserializer<'de>,
+    {
+        let bytes = Vec::deserialize(deserializer)?;
+        PedersenCommitment::from_bytes(&bytes[..]).map_err(|e| de::Error::custom(e.to_string()))
+    }
+}
+
+impl PartialEq for PedersenCommitment {
+    fn eq(&self, other: &Self) -> bool {
+        self.point == other.point
+    }
+}
+
+impl Eq for PedersenCommitment {}
+
+impl PartialOrd for PedersenCommitment {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        self.as_ref().partial_cmp(other.as_ref())
+    }
+}
+
+impl Ord for PedersenCommitment {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.as_bytes().cmp(other.as_bytes())
+    }
+}
+
+///
+/// Bulletproof Range Proofs
+///
+
+#[derive(Debug)]
+pub struct BulletproofsRangeProof {
+    proof: RangeProof,
+    bytes: OnceCell<Vec<u8>>,
+}
+
+impl BulletproofsRangeProof {
+    /// Prove that commited value is within the range [0, 2^64). Function only works
+    /// for n = 8, 16, 32, 64
+    pub fn prove_bound(
+        value: u64,
+        blinding: [u8; 32],
+        n: usize,
+        domain: &'static [u8]
+    ) -> Result<(PedersenCommitment, Self), signature::Error> {
+        // Although this is also checked in the bulletproofs library, we check again
+        // to avoid unexpected behaviour in the case of library updates
+        if !(n == 8 || n == 16 || n == 32 || n == 64) {
+            return Err(signature::Error::new());
```

Comment:
Better error msg? No worries just add a TODO and we can add in an upcoming PR for all errors in this commit